whylog/constraints/__init__.py

```diff
@@ -2,42 +2,180 @@
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstructorGroupsError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    TYPE = 'undefined'
+
+    MIN_GROUPS_COUNT = None
+    MAX_GROUPS_COUNT = None
+
+    PARAMS = []
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def __init__(self, param_dict, groups):
```

Comment:
My proposition is to use groups as the first parametr, and the param_dict as the second with None value by the default. If somtimes it can be empty dict I don't see why it should be the first param of the constructor.